examples/seq2seq/test_finetune_trainer.py

```diff
@@ -195,12 +218,17 @@ def run_trainer(self, eval_steps: int, max_len: str, model_name: str, num_train_
 
         n_gpu = get_gpu_count()
         if n_gpu > 1:
-            distributed_args = f"""
-                -m torch.distributed.launch
-                --nproc_per_node={n_gpu}
-                {self.test_file_dir}/finetune_trainer.py
-            """.split()
-            cmd = [sys.executable] + distributed_args + args
+            if dist is None:
+                assert f"for n_gpu={n_gpu} pass `dp` or `dpp` for the `dist` arg"
+            launch_args = []
+            if dist == "ddp":
+                launch_args = f"""
+                    -m torch.distributed.launch
+                    --nproc_per_node={n_gpu}
+                """.split()
+            launch_args.append(f"{self.test_file_dir}/finetune_trainer.py")
```

Comment:
If `dist=dp` the `n_gpu` argument is not used here, so maybe this should be in a separate test?

Reply:
Sorry, I don't follow your question - could you please expand on it?

Currently `dist=dp` runs here, it is just `dist=dpp` is a special case where it pushes extra args and hence needs `--nproc_per_node={n_gpu}`

Under `dist=dp` the `n_gpu` is sorted out inside trainer, is it not? or do you suggest it should have been passed as an argument to the trainer?

Reply:
Using `dist='dp'` does not require the `n_gpu` argument, yet we only launch it inside a test regarding that `n_gpu`, which I find weird.
I'm not familiar with this part of the code so maybe I'm missing something here.

Reply:
since this is wrapper able to run under any setup,  it uses `n_gpu` to know whether to:

1. launch a distributed test (dp or dpp)
2. pass `--nproc_per_node={n_gpu}` in case of `dpp`

We could explicitly pass `n_gpu` in case of `dist=dp` if it makes things more clear.

Also perhaps it's not clear how `dist=dp` fits into this picture so may be the following would be a more readable code:
```
        n_gpu = get_gpu_count()
        if n_gpu > 1:
            if dist == "dp":
                launch_args = []
            elif dist == "ddp":
                launch_args = f"""
                    -m torch.distributed.launch
                    --nproc_per_node={n_gpu}
                """.split()
            else:
                assert f"for n_gpu={n_gpu} pass `dp` or `dpp` for the `dist` arg"
            launch_args.append(f"{self.test_file_dir}/finetune_trainer.py")
```
```

Reply:
I guess I don't understand why you need to make a special case for dp compared to the regular script. Launching a script in dp is the same as launching it period. Pytest will use dp if you have several GPUs and the tests you are running use it (always have to set a `CUDA_VISIBLE_DEVICES="1"` on my machine to avoid that for instance).

Reply:
Ah, I finally see what you mean - thank you for clarifying, @sgugger. Let me experiment and get back to you.

Reply:
OK, sorted it out - simplified and re-tested. Thank you for that dialog, @sgugger - a much better outcome thanks to you.